usb-c: Refactor USB-C Subsystem Sink

sambhurst opened the pull request:
Refactor USB-C Subsystem Sink so that Power Delivery Source support and be easily added.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb-c: Refactor USB-C Subsystem Sink

Refactor USB-C Subsystem Sink so that Power Delivery
Source support and be easily added.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

alevkoy commented on subsys/usb/usb_c/usbc_tc_common.h:
```diff
@@ -45,6 +75,30 @@ struct tc_sm_t {
+ * @brief Get current TypeC State
+ *
+ * @param dev Pointer to the device structure for the driver instance
+ * @return current TypeC state
```
Nit: Here and elsewhere, "Type-C"

sambhurst replied:
done

alevkoy commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -1583,7 +882,7 @@ static void pe_drs_evaluate_swap_run(void *obj)
  * @brief PE_DRS_Send_Swap Entry state
  */
-static void pe_drs_send_swap_entry(void *obj)
+void pe_drs_send_swap_entry(void *obj)
```
A lot of these states, for instance this one, are not sink-specific. Should they be in the common file?

sambhurst replied:
This was done in a later PR but should be done here. I've moved it to common

sambhurst replied:
I've removed the functions

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.h:
```diff
@@ -218,4 +320,120 @@ enum tc_power_role pe_get_power_role(const struct device *dev);
+/**
+ * @brief Handle sink-specific DPM requests
+ */
+bool sink_dpm_requests(const struct device *dev);
```
Should this be in the sink header file?

sambhurst replied:
This was changed in a later PR but should be done here. I've removed it.

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.h:
```diff
@@ -14,6 +14,12 @@
+ * @brief The HardResetCounter is used to retry the Hard Reset whenever there
+ *        is no response from the remote device (see Section 6.6.6)
+ */
+#define N_HARD_RESET_COUNT 2
```
Do we have a header file somewhere with a bunch of counters/timeout values, etc. from the specs collected in one place?

sambhurst replied:
This was added in a later PR [53275](https://github.com/zephyrproject-rtos/zephyr/pull/53275)

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.h:
```diff
@@ -22,6 +28,102 @@ enum pe_error {
+/**
+ * @brief Policy Engine Layer States
+ */
+enum usbc_pe_state {
```
Up to this point, things like state IDs and flags were contained in the state machine .c files, and that meant that only that state machine could access them by definition. Now files from other state machines are including this one, and that might make someone think that it's OK to touch this stuff from outside the PE. I think it might be safer to have something like `usbc_pe_*_internal.h` for things that need to be shared among PE files but are not part of the PE's public API.

Same concern for the TC.

sambhurst replied:
Done

alevkoy replied:
I guess there are multiple levels of internal. Except for `usbc_stack.h`, all of this stuff is internal to the PD stack, but the things I was talking about are further internal to the PE or the TC, and this change elevates their visibility to broadly within the PD stack.

What I had in mind here is that the public API of the PE would remain in something like `usbc_pe.h`, and files in the TC or the PRL or the USB-C task would include `usbc_pe.h`. There would be a separate file like `usbc_pe_common_internal.h`, and only other files in the PE, like `usbc_pe_common.c` or `usbc_pe_sink.c`, would include `usbc_pe_common_internal.h`. I wasn't thinking of changing the name of the sink states header file to include internal as you just did, but I like it.

Same deal with the TC.

sambhurst replied:
I've updated the code. Thanks

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,718 @@
+ * @brief Policy engine State table
+ */
+static const struct smf_state pe_states[] = {
+	[PE_SNK_STARTUP] = SMF_CREATE_STATE(
```
Right now, even though the state machine is split up, the two parts can't exist without each other. The structures of the state machine (constant state array, single state enum) mean that the C files can't be completely separated, and even if they could, there's no such thing as a port that is neither Source nor Sink.

I don't immediately see what this refactor buys us, but I'm guessing that the advantages will become more obvious when I review the source support PR.

sambhurst replied:
I thought it would be easier to review and maintain if the each file contained similar functionality, instead of having everything in one large file.

alevkoy replied:
In the long term, will all the files that make up the PE still get compiled together, or will some of them be left out based on the config?

sambhurst replied:
Yes. In a source only device, the pe_snk_* files will not be compiled in, and vice versa.

sambhurst replied:
Yes

keith-zephyr commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,848 @@
+{
+	struct usbc_port_data *data = dev->data;
+
+	smf_set_state(SMF_CTX(data->pe), &pe_states[state]);
```
This need a bounds check that `state < ARRAY_SIZE(pe_states)`.  You could also wrap this in a macro to do a build time assert (see the `tc_set_state()` example).

sambhurst replied:
Same error with ARRAY_SIZE(pe_states)

keith-zephyr replied:
Is the build error from the run-time check or the build-time check?  `__ASSERT(state < ARRAY_SIZE(pe_states), "invalid pe_state %d", state);` should work.

sambhurst replied:
Yes that worked. I was using BUILD_ASSERT

keith-zephyr commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,848 @@
+/**
+ * @brief Run the Policy Engine layer
+ */
+void pe_run(const struct device *dev,
+	    const int32_t dpm_request)
```
Please run clang-format on this file.

sambhurst replied:
Done

keith-zephyr commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,848 @@
+		pe_suspend_run,
+		NULL,
+		NULL),
+};
```
This needs a build time check that `ARRAY_SIZE(pe_states) == PE_STATE_COUNT`.  Where `PE_STATE_COUNT = PE_SUSPEND+1`.

sambhurst replied:
Done

keith-zephyr commented on subsys/usb/usb_c/usbc_pe_common_internal.h:
```diff
@@ -22,6 +22,102 @@ enum pe_error {
+/**
+ * @brief Policy Engine Layer Flags
+ */
+enum pe_flags {
```
Note - in `struct policy_engine` defined in usbc_pe.h you declare flags as `atomic_t flags`.  This is usually limited to 32-bits.

You should either assert that `enum pe_flags` fits in 32-bits, or you need to update `struct policy_engine`:

```
struct policy_engine {
  ATOMIC_DEFINE(flags, PE_FLAGS_COUNT);
```

Zephyr's atomic support automatically supports more than 32-bit when using ATOMIC_DEFINE.

sambhurst replied:
Done

keith-zephyr commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -0,0 +1,264 @@
+	struct usbc_port_data *data = dev->data;
+	struct tc_sm_t *tc = data->tc;
+
+	smf_set_state(SMF_CTX(tc), &tc_states[state]);
```
Check that `state < ARRAY_SIZE(tc_states)`.  This could potentially be a build time check because all callers to `tc_set_state()` use a build time constant.

```
#define TC_SET_STATE(dev, state) \
    BUILD_ASSERT(state < ARRAY_SIZE(tc_states)); \
    smf_set_state(SMF_CTX(tc), &tc_states[state]);
```

sambhurst replied:
ARRAY_SIZE(tc_states) causes an error

sambhurst replied:
__ASSERT works but BUILD_ASSERT causes the error

keith-zephyr commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -0,0 +1,264 @@
+/**
+ * @brief Enable Power Delivery
+ */
+void pd_enable(const struct device *dev,
```
This public routine doesn't follow the naming convention of the other TC function.  This should either get moved to another module or should use the `tc_` prefix.

sambhurst replied:
done

DianaZig commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -0,0 +1,839 @@
+	 * Transition to the PE_SNK_Transition_to_default state when:
+	 *  1) The Hard Reset is complete.
+	 */
+	if (atomic_test_bit(&pe->flags, PE_FLAGS_HARD_RESET_PENDING)) {
```
I think I missed where this bit gets set in the first place?

sambhurst replied:
I added code to set the bit in the entry function

DianaZig commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -0,0 +1,839 @@
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+
+	/* Wait until message is sent or dropped */
```
This is actually a state they changed recently in PD 3.1.  It used to be "send the get_source_cap, and just head back to ready", but now it runs the sender response timer and follows some trickily written "if" checks to either snk ready or snk_evaluate_capability

sambhurst replied:
I've updated the state

DianaZig commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,848 @@
+		} else {
+			/* Protocol Error */
+			policy_notify(dev, PROTOCOL_ERROR);
+			pe_send_soft_reset(dev, PD_PACKET_SOP);
```
And now I'm just going to go ahead and add onto this once again...

Reading 8.3.3.4.2.1 PE_SNK_Send_Soft_Reset State while I looked at the soft reset state, apparently we need to perform ErrorRecovery when we hit issues during a DRS.  That one I'd never noticed before, but apparently it's been in the spec since at least 3.0/1.2

sambhurst replied:
I've updated the state to perform Error Recovery

DianaZig commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -0,0 +1,839 @@
+		if (atomic_test_bit(&pe->flags, PE_FLAGS_PROTOCOL_ERROR)) {
+			atomic_clear_bit(&pe->flags, PE_FLAGS_PROTOCOL_ERROR);
+		} else {
+			policy_notify(dev, PORT_PARTNER_NOT_RESPONSIVE);
```
Are we using this notification for more than just the partner not sending PD messages in general then?  What's it's stated purpose/meaning?

sambhurst replied:
I removed the policy_notify

DianaZig commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -0,0 +1,839 @@
+
+	if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_TX_COMPLETE)) {
+		pe_set_state(dev, PE_SNK_WAIT_FOR_CAPABILITIES);
+	} else if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_PROTOCOL_ERROR)) {
```
Out of curiosity... I've seen this bit referenced a bit, but not sure who sets it? 

Also, hard reset if our Accept was discarded?

sambhurst replied:
Good catch. I forgot to remove these PE_FLAGS_PROTOCOL_ERROR flag checks. They have been removed

DianaZig commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -0,0 +1,839 @@
+	if (usbc_timer_expired(&pe->pd_t_chunking_not_supported)) {
+		pe_set_state(dev, PE_SEND_NOT_SUPPORTED);
+	}
+}
```
I'm not super familiar with zephyr's timers, but does this one need stopped on exit?

sambhurst replied:
I added the exit state that stops the timer

DianaZig commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -0,0 +1,264 @@
+	 * Start out in error recovery state so the CC lines are opened for a
+	 * short while if this is a system reset.
+	 */
+	tc_set_state(dev, TC_ERROR_RECOVERY_STATE);
```
Should the initial state be a DPM callback, or is that a feature we'll add later?

sambhurst replied:
Yes. Let's add this feature later.

sambhurst committed:
usb_c: run clang-format

Run clang-format

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Use atomic arrays for PE flags

Use atomic arrays for PE flags because the total number
of Policy Engine flags could exceed 32

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Check state machine array of states

Build time check that the number of states initialized
in the state machine array matches the number of enums
used to index said array.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Fix DPM request handlers

Processing the common DPM request handler in the
Sink DPM request handler.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Properly handle Protocol Errors during DRS

From 8.3.3.4.2.1 PE_SNK_Send_Soft_Reset State, Error Recovery
should be performed if a Protocol Error happens during a Data
Role Swap.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Transition to Sink Ready on sender response timeout

Transition to pe_snk_ready state from pe_drs_send_swap_run,
on sender response timeout

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Remove unnecessary code from PE_Send_Soft_Reset

Remove unnecessary code from PE_Send_Soft_Reset

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Remove unnecessary PE_FLAGS_PROTOCOL_ERROR flag

Remove unnecessary PE_FLAGS_PROTOCOL_ERROR flag

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Stop chunking_not_supported timer on exit

Stop chunking_not_supported timer on exit

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,848 @@
+	if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+		/* Protocol Error */
+		policy_notify(dev, PROTOCOL_ERROR);
+		pe_send_soft_reset(dev, PD_PACKET_SOP);
```
Done

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,848 @@
+		} else {
+			/* Pass the DPM request along to the state machine */
+			pe->dpm_request = dpm_request;
+			common_dpm_requests(dev);
```
Good catch. I've moved the comm DPM request handler after state machine runs.

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -0,0 +1,848 @@
+			pe_set_state(dev, PE_DRS_SEND_SWAP);
+		} else if (pe->dpm_request == REQUEST_PE_SOFT_RESET_SEND) {
+			pe_set_state(dev, PE_SEND_SOFT_RESET);
+		}
```
I've changed the code so the the Source/Sink DPM request handler is run before the common DPM request handler. So any request not handled by the Source/Sink DPM request handler will be passed to the common DPM request handler.

sambhurst commented on subsys/usb/usb_c/CMakeLists.txt:
```diff
@@ -2,4 +2,8 @@
 
-zephyr_library_sources_ifdef(CONFIG_USBC_STACK usbc_timer.c usbc_stack.c usbc_tc.c usbc_prl.c usbc_pe.c)
+# Common USB-C Stack files
+zephyr_library_sources_ifdef(CONFIG_USBC_STACK usbc_timer.c usbc_stack.c usbc_tc_common.c usbc_pe_common.c usbc_prl.c)
```
Done

sambhurst commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -38,6 +38,14 @@ config BUILD_OUTPUT_VIF
+
+config USBC_SNK_CSM
+	bool "Sink USBC Connection State Machine"
+
```
Done

keith-zephyr replied:
Nit - why is sink capitalized?  Also, specify that this option configures the CSM for sink operation only.  Because this is a choice symbol I assume that you will also eventually add a DRP option.

What do you think about naming this option `USBC_CSM_SINK_ONLY`?

alevkoy replied:
"Sink" and "Source" are proper nouns in the Type-C spec.

sambhurst replied:
I changed the name

sambhurst commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -38,6 +38,14 @@ config BUILD_OUTPUT_VIF
 	depends on DT_HAS_USB_C_CONNECTOR_ENABLED
 	help
 	  Generate XML file containing VIF policies during project build.
+choice
```
Done

sambhurst commented on subsys/usb/usb_c/usbc_pe_common_internal.h:
```diff
@@ -218,4 +314,108 @@ enum tc_power_role pe_get_power_role(const struct device *dev);
+
+/**
+ * @brief Notify the DPM of a WAIT message reception
+ */
```
Done

sambhurst committed:
samples: subsys: usb_c: sink: Add missing policy notifications

Add missing policy notifications to case statement

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -151,14 +149,19 @@ void pe_run(const struct device *dev, const int32_t dpm_request)
+		 * Sink Ready state
+		 */
+		if (pe_get_state(dev) == PE_SNK_READY) {
+			common_dpm_requests(dev);
```
Would it make more sense to just call this from the ready state? That would make this function care less about what the PE state is.

sambhurst replied:
I was calling common_dpm_requests here because it was common to both Source and Sink but I can just call the function from both ready states as you suggested.

alevkoy commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -29,6 +29,14 @@ void tc_subsys_init(const struct device *dev)
+ */
+void tc_start_error_recovery(const struct device *dev)
+{
+	tc_set_state(dev, TC_ERROR_RECOVERY_STATE);
```
I can't remember what our naming convention is for internal vs API functions for the individual state machines, but is it safe to change the TC state asynchronously from the PE?

sambhurst replied:
I agree that this could be a problem and I'v changed the code.

alevkoy replied:
I can't remember exactly how this code was before, but it looks like `pe_drs_send_swap_run` is still going to indirectly call `tc_set_state` while the TC is not running. Downstream, I think we'd typically have `tc_start_error_recovery` set a flag that the TC checked on the next iteration.

sambhurst replied:
I messed up my rebasing and merged the change in the following commit. I'll fix it and upload another PR

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -703,10 +703,9 @@ void pe_drs_send_swap_run(void *obj)
 		return;
 	}
 
+	/* Transition to PE_SNK_READY on SenderResponseTimer timeout */
```
Nit: This commit message should say in what PE state this change is being made.

sambhurst replied:
The commit message has been updated

alevkoy commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -109,6 +109,10 @@ enum usbc_policy_notify_t {
 	POWER_CHANGE_1A5,
 	/** Sink SubPower state a 5V / 3A */
 	POWER_CHANGE_3A0,
+	/** Sender Response Timeout */
```
Do these belong in the "missing policy notifications" commit?

sambhurst replied:
The messing policy's commit is for the sample code, so I think it better to keep these in this PR.

alevkoy commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -607,12 +612,40 @@ void pe_snk_get_source_cap_run(void *obj)
+			 * Inform the DPM of the reception of the source
+			 * capabilities
+			 */
+			policy_notify(dev, SOURCE_CAPABILITIES_RECEIVED);
```
It seems more foolproof to just do this unconditionally in `pe_snk_evaluate_capability_entry`.

Also, the arrangement here breaks the "return immediately after `pe_set_state` rule. I don't think that would actually be a problem here, but it's probably easier if we keep it absolute.

sambhurst replied:
I didn't want to send the notification until the Source Caps message was received and I switched the order of the notification and state change.

alevkoy replied:
I'm OK with this as is, but wouldn't calling `policy_notify` from `pe_snk_evaluate_capability_entry` still guarantee that the notification doesn't send until the Source Caps have been received?

sambhurst replied:
I guess I needed to read your suggestion twice. I've moved the notification to pe_snk_evaluate_capability_entry.

alevkoy commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -702,19 +702,26 @@ void pe_send_soft_reset_run(void *obj)
-		if (atomic_test_bit(pe->flags, PE_FLAGS_PROTOCOL_ERROR)) {
-			atomic_clear_bit(pe->flags, PE_FLAGS_PROTOCOL_ERROR);
-		} else {
-			policy_notify(dev, PORT_PARTNER_NOT_RESPONSIVE);
```
Should you add an additional check to the beginning of `pe_snk_hard_reset_entry` to send this notification, or was it being sent in error in this removed case?

sambhurst replied:
The notification is already sent in the pe_snk_hard_reset_entry.

sambhurst commented on subsys/usb/usb_c/usbc_pe_common_internal.h:
```diff
@@ -218,4 +314,108 @@ enum tc_power_role pe_get_power_role(const struct device *dev);
+/**
+ * @brief Send a Power Delivery Data Message
+ */
+void send_data_msg(const struct device *dev,
```
I think we're on the honor system where no function defined in pe_x_internal.h should be called directly by another layer. I added pe_ to send_data_msg and send_ctrl_msg

sambhurst commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -0,0 +1,839 @@
+	 */
+	if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_MSG_RECEIVED)) {
+		header = prl_rx->emsg.header;
+		if ((header.extended == false) &&
```
yes and I've removed the check here

alevkoy reviewed (commented):
I'm still wondering whether it's safe to change the TC state from inside the PE. Otherwise, this looks fine.
